### PR TITLE
python3Packages.sourmash: init at 4.8.3

### PR DESCRIPTION
**Pull request by luizirber**
###### Description of changes

`sourmash` is a bioinformatics tool for quickly searching, comparing, and analyzing genomic and metagenomic data sets. 

Homepage: https://sourmash.bio
GitHub: https://github.com/sourmash-bio/sourmash

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [-] (Package updates) Added a release notes entry if the change is major or breaking
  - [-] (Module updates) Added a release notes entry if the change is significant
  - [-] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/sourmash/default.nix [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    name = "${pname}-${version}";
+    hash = "sha256-/BPkZ/EtB8J3s72G8ADSZHrKFi5PRNexWsboNmM+QwA=";
+  };
+
+  nativeBuildInputs = with rustPlatform; [ cargoSetupHook maturinBuildHook ];
+  propagatedBuildInputs = [ bitstring cachetools cffi deprecation matplotlib numpy scipy screed ];
+
+  pythonImportsCheck = [ "sourmash" ];
+  nativeCheckInputs = [ pytestCheckHook ];
+  checkInputs = [ hypothesis pytest-xdist pyyaml ];
```
[the remark this answers is not in the document]
Comment (luizirber) [reply pkgs/development/python-modules/sourmash/default.nix]: Done, thanks!

**Comment by SuperSandro2000** [comment]
@ofborg build

File: pkgs/development/python-modules/sourmash/default.nix [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  };
+
+  cargoDeps = rustPlatform.fetchCargoTarball {
+    inherit src;
+    name = "${pname}-${version}";
+    hash = "sha256-/BPkZ/EtB8J3s72G8ADSZHrKFi5PRNexWsboNmM+QwA=";
+  };
+
+  nativeBuildInputs = with rustPlatform; [ cargoSetupHook maturinBuildHook ];
+  propagatedBuildInputs = [ bitstring cachetools cffi deprecation matplotlib numpy scipy screed ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/sourmash/default.nix:38]:
```suggestion
  nativeBuildInputs = with rustPlatform; [
    cargoSetupHook
    maturinBuildHook
  ];

  buildInputs = [ iconv ];

  propagatedBuildInputs = [
    bitstring
    cachetools
    cffi
    deprecation
    matplotlib
    numpy
    scipy
    screed
  ];
```
we normally format things like this in python packages to increase readability. Also iconv should fix the darwin compile error. On linux it is null.

**Comment by SuperSandro2000** [comment]
one test fails:


```
=================================== FAILURES ===================================
_ test_gather_check_scaled_bounds_less_than_minimum[--no-prefetch---no-linear] _
[gw58] linux -- Python 3.10.11 /nix/store/95cxzy2hpizr23343b8bskl4yacf4b3l-python3-3.10.11/bin/python3.10
runtmp = <sourmash_tst_utils.RunnerContext object at 0x7ffbe29cd0f0>
prefetch_gather = '--no-prefetch', linear_gather = '--no-linear'
    def test_gather_check_scaled_bounds_less_than_minimum(runtmp, prefetch_gather, linear_gather):
        testdata_glob = utils.get_test_data('gather/GCF*.sig')
        testdata_sigs = glob.glob(testdata_glob)
    
        query_sig = utils.get_test_data('gather/combined.sig')
    
        with pytest.raises(SourmashCommandFailed):
            runtmp.sourmash('gather', query_sig, prefetch_gather, linear_gather, 'gcf_all', '-k', '21', '--scaled', '50', '--threshold-bp', '50000')
    
>       assert "WARNING: scaled value should be >= 100. Continuing anyway." in runtmp.last_result.err
E       assert 'WARNING: scaled value should be >= 100. Continuing anyway.' in '\r\x1b[Kselecting specified query k=21\n\r\x1b[Kloaded query: -... (k=21, DNA)\n\r\x1b[Kdownsampling query from scale...path.py", line 76, in join\n    a = os.fspath(a)\nTypeError: expected str, bytes or os.PathLike object, not NoneType\n'
E        +  where '\r\x1b[Kselecting specified query k=21\n\r\x1b[Kloaded query: -... (k=21, DNA)\n\r\x1b[Kdownsampling query from scale...path.py", line 76, in join\n    a = os.fspath(a)\nTypeError: expected str, bytes or os.PathLike object, not NoneType\n' = ScriptResults(status=-1, out='', err='\r\x1b[Kselecting specified query k=21\n\r\x1b[Kloaded query: -... (k=21, DNA)\n...ath.py", line 76, in join\n    a = os.fspath(a)\nTypeError: expected str, bytes or os.PathLike object, not NoneType\n').err
E        +    where ScriptResults(status=-1, out='', err='\r\x1b[Kselecting specified query k=21\n\r\x1b[Kloaded query: -... (k=21, DNA)\n...ath.py", line 76, in join\n    a = os.fspath(a)\nTypeError: expected str, bytes or os.PathLike object, not NoneType\n') = <sourmash_tst_utils.RunnerContext object at 0x7ffbe29cd0f0>.last_result
tests/test_sourmash.py:4417: AssertionError
----------------------------- Captured stdout call -----------------------------
running: sourmash in: /build/sourmashtest_33f6e022
arguments ['sourmash', 'gather', '/build/sourmash-4.8.2/tests/test-data/gather/combined.sig', '--no-prefetch', '--no-linear', 'gcf_all', '-k', '21', '--scaled', '50', '--threshold-bp', '50000']
=============================== warnings summary ===============================
```

**Comment by doronbehar** [comment]
Result of `nixpkgs-review pr 230596` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.sourmash</li>
    <li>python310Packages.sourmash.dist</li>
    <li>python311Packages.sourmash</li>
    <li>python311Packages.sourmash.dist</li>
  </ul>
</details>

**Comment by luizirber** [comment]
I'm very confused with the current build error... Before it was related to tests, but now it is because a weird mode `130` is being inserted in the [`ffi.dlopen`](https://cffi.readthedocs.io/en/latest/cdef.html#ffi-dlopen-loading-libraries-in-abi-mode), but that doesn't happen by regular Python packaging tooling (nor by running `nix build github:sourmash-bio/sourmash# ; cat result/lib/python3.10/site-packages/sourmash/_lowlevel/__init__.py`, the flake-based package building I have in the upstream repo).

```
$ nix build .#python311Packages.sourmash
error: builder for '/nix/store/ha2j8l482pn0va7wkddb96vasvdmab4a-python3.11-sourmash-4.8.2.drv' failed with exit code 1;
       last 10 log lines:
       >   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
       >   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
       >   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
       >   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
       >   File "/nix/store/qgv5ldklwg064vzdal7djzc1wv85f2l2-python3.11-sourmash-4.8.2/lib/python3.11/site-packages/sourmash/__init__.py", line 33, in <module>
       >     from ._lowlevel import ffi, lib
       >   File "/nix/store/qgv5ldklwg064vzdal7djzc1wv85f2l2-python3.11-sourmash-4.8.2/lib/python3.11/site-packages/sourmash/_lowlevel.py", line 7, in <module>
       >     lib = ffi.dlopen(os.path.join(os.path.dirname(__file__), '_lowlevel__lib.so'), 130)
       >           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       > OSError: cannot load library '/nix/store/qgv5ldklwg064vzdal7djzc1wv85f2l2-python3.11-sourmash-4.8.2/lib/python3.11/site-packages/sourmash/_lowlevel__lib.so': invalid mode parameter
       For full logs, run 'nix log /nix/store/ha2j8l482pn0va7wkddb96vasvdmab4a-python3.11-sourmash-4.8.2.drv'.
```

Is anyone aware of a Python hook that could be changing that? Not sure how to fix this...

**Review by natsukium — COMMENTED** [review]
While this PR is open, `maturin` of nixpkgs is reached 1.0.0 in #230799.
This is why the build failed.
I can build it by applying these two patches.
https://github.com/sourmash-bio/sourmash/commit/9de27b6601af209ea69696c21586e755a5ca72c6
https://github.com/sourmash-bio/sourmash/commit/406b94598ec3660408e9d71610c7283e22f81ddf

**Comment by luizirber** [comment]
> While this PR is open, `maturin` of nixpkgs is reached 1.0.0 in #230799. This is why the build failed. I can build it by applying these two patches. [sourmash-bio/sourmash@9de27b6](https://github.com/sourmash-bio/sourmash/commit/9de27b6601af209ea69696c21586e755a5ca72c6) [sourmash-bio/sourmash@406b945](https://github.com/sourmash-bio/sourmash/commit/406b94598ec3660408e9d71610c7283e22f81ddf)

Indeed, thanks @natsukium! I will do a new release of the upstream project and update the PyPI download link

File: pkgs/development/python-modules/sourmash/default.nix [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+buildPythonPackage rec {
+  pname = "sourmash";
+  version = "4.8.3";
+  format = "pyproject";
+  disabled = pythonOlder "3.8";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-LIMpL9cLafytRFyPam/FBNi757j1v6o1FG/K2JknDQY=";
```
Comment (natsukium) [inline pkgs/development/python-modules/sourmash/default.nix:29]:
```suggestion
    hash = "sha256-LIMpL9cLafytRFyPam/FBNi757j1v6o1FG/K2JknDQY=";
```

File: pkgs/development/python-modules/sourmash/default.nix [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  # TODO(luizirber): Working on fixing these upstream
+  disabledTests = [
+    "test_compare_no_such_file"
+    "test_do_sourmash_index_multiscaled_rescale_fail"
+    "test_metagenome_kreport_out_fail"
+  ];
+
+  meta = with lib; {
+    description = "Quickly search, compare, and analyze genomic and metagenomic data sets";
+    homepage = "https://sourmash.bio";
```
Comment (natsukium) [inline pkgs/development/python-modules/sourmash/default.nix:73]:
```suggestion
    homepage = "https://sourmash.bio";
    changelog = "https://github.com/sourmash-bio/sourmash/releases/tag/v${version}";
```
Adding the `meta.changelog` is helpful for future updates.

**Comment by natsukium** [comment]
@ofborg build python310Packages.sourmash

File: pkgs/development/python-modules/sourmash/default.nix (hunk 4 above)
Comment (luizirber) [reply pkgs/development/python-modules/sourmash/default.nix:73]: fixed!

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 230596` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.sourmash</li>
    <li>python310Packages.sourmash.dist</li>
    <li>python311Packages.sourmash</li>
    <li>python311Packages.sourmash.dist</li>
  </ul>
</details>

**Review by natsukium — APPROVED** [review]
LGTM

**Pull request merged**